python/pyspark/sql/dataframe.py

```diff
@@ -674,7 +674,7 @@ def cache(self):
         .. note:: The default storage level has changed to `MEMORY_AND_DISK` to match Scala in 2.0.
         """
         self.is_cached = True
-        self._jdf.cache()
+        self.persist(StorageLevel.MEMORY_AND_DISK)
```

Comment:
DataFrame itself is not serialized via Python unless you want to collect/createDataFrame/Python UDFs way whereas Python RDD are serialized via Python pickle. The code paths should be same as Scala's. I don't think this change is necessary.

Reply:
yea, no matter you call `df.cache` in Scala or Python, the final RDD to cache is `RDD[CachedBatch]` not `RDD[Array[Byte]]`. I don't think SPARK-2014 applies here.

Reply:
Okay you might be right here but I observed a difference in the behavior of `pySparkDataframe.cache()` and `pySparkDataframe.persist()` because of differences in Storage Levels. When `cache()` is used, RDDs are spilled to disk but when `persist()` is used RDDs are persisted in memory. 

This is because `MemoryStore#putIteratorAsValues()` is used when `cache()` is called and `MemoryStore#putIteratorAsBytes()` is used when `persist()` is called due to difference in storage levels.

My example:
```
df = spark.range(1,4500000000).cache()
df.count()
AND
df = spark.range(1,4500000000).persist()
df.count()
```

I tried this on Spark version 2.4.6. I have attached RDD storage screenshots here. Correct me if I'm missing something and this is expected.
 
![image](https://user-images.githubusercontent.com/43843989/88503802-69105480-cff0-11ea-9609-ba1e7362b5b2.png)


 
![image](https://user-images.githubusercontent.com/43843989/88503694-1f276e80-cff0-11ea-8b60-6f8bcd6eafd6.png)

Reply:
Does the same thing happen with Scala API too?

Reply:
No for scala behavior is the same whether `cache()` or `persist()` is used because storage level used is same in both the cases.

Reply:
Isn't it true that every where(except Dataframe.py) whether scala or python APIs, persist means `MEMORY_ONLY` ?

Is there a specific reason to change that?

Reply:
No in `Dataset.scala` persist means persisting with default storage level `MEMORY_AND_DISK`
```
/**
   * Persist this Dataset with the default storage level (`MEMORY_AND_DISK`).
   *
   * @group basic
   * @since 1.6.0
   */
  def persist(): this.type = {
    sparkSession.sharedState.cacheManager.cacheQuery(this)
    this
  }

  /**
   * Persist this Dataset with the default storage level (`MEMORY_AND_DISK`).
   *
   * @group basic
   * @since 1.6.0
   */
  def cache(): this.type = persist()
```

Reply:
> Okay you might be right here but I observed a difference in the behavior of `pySparkDataframe.cache()` and `pySparkDataframe.persist()` because of differences in Storage Levels. When `cache()` is used, RDDs are spilled to disk but when `persist()` is used RDDs are persisted in memory.
> 
> This is because `MemoryStore#putIteratorAsValues()` is used when `cache()` is called and `MemoryStore#putIteratorAsBytes()` is used when `persist()` is called due to difference in storage levels.
> 
> My example:
> 
> ```
> df = spark.range(1,4500000000).cache()
> df.count()
> AND
> df = spark.range(1,4500000000).persist()
> df.count()
> ```
> 
> I tried this on Spark version 2.4.6. I have attached RDD storage screenshots here. Correct me if I'm missing something and this is expected.
> 
> ![image](https://user-images.githubusercontent.com/43843989/88503802-69105480-cff0-11ea-9609-ba1e7362b5b2.png)
> 
> ![image](https://user-images.githubusercontent.com/43843989/88503694-1f276e80-cff0-11ea-8b60-6f8bcd6eafd6.png)

@HyukjinKwon 
Can you please comment if the approach taken by this PR to solve this bug is correct or not? If it's not correct, can you please create a new JIRA and explain the issue and I can close this PR.
cc @srowen @cloud-fan @ScrapCodes

Reply:
Hm, yeah that is strange. What if Pyspark cache() just calls persist() in Pyspark, rather than cache() in Scala?

Reply:
@srowen 
Do you mean changing `self.persist(StorageLevel.MEMORY_AND_DISK)` to `self.persist()` at line 677? I believe the effect will be the same because the default storage level for Pyspark persist is also MEMORY_AND_DISK. 

```
@since(1.3)
    def persist(self, storageLevel=StorageLevel.MEMORY_AND_DISK):
        """Sets the storage level to persist the contents of the :class:`DataFrame` across
        operations after the first time it is computed. This can only be used to assign
        a new storage level if the :class:`DataFrame` does not have a storage level set yet.
        If no storage level is specified defaults to (`MEMORY_AND_DISK`).
        .. note:: The default storage level has changed to `MEMORY_AND_DISK` to match Scala in 2.0.
        """
        self.is_cached = True
        javaStorageLevel = self._sc._getJavaStorageLevel(storageLevel)
        self._jdf.persist(javaStorageLevel)
        return self
```

Reply:
Yes but that's the desired level, no? I understand there is a subtle difference between the meaning of that in Scala and Python, but, would that be the difference you are trying to fix here?

Reply:
using Pyspark's `self.persist(StorageLevel.MEMORY_AND_DISK)` or Pyspark's `self.persist()` will have the same impact since in both the cases Python's default storage level will be used i.e. `StorageLevel.MEMORY_AND_DISK = StorageLevel(True, True, False, False)` with `deserialized=false`. Hence both changes are equivalent and either will fix the bug.

Note that calling `self._sc._getJavaStorageLevel(StorageLevel.MEMORY_AND_DISK)` in pyspark doesn't return scala's default storage level. Instead, it returns `new StorageLevel(true, true, false, false)` which has `deserialized=false` and hence it will fix the bug.

However, in current master since scala's cache is called `self._jdf.cache()`, storage level uses is scala's default ie. `val MEMORY_AND_DISK = new StorageLevel(true, true, false, true)` which has `deserialized=true` leading to the bug.

Reply:
OK, right. That could be a simpler fix. But then I'm still not so much following why there is a behavior difference per https://github.com/apache/spark/pull/29242#discussion_r462086489 I'm trying to understand whether this difference is expected or not, per earlier comments.

Reply:
@srowen
The difference in behavior is due to this code in `BlockManager.scala`.  
```
// Put it in memory first, even if it also has useDisk set to true;
        // We will drop it to disk later if the memory store can't hold it.
        if (level.deserialized) {
          memoryStore.putIteratorAsValues(blockId, iterator(), classTag) match {
            case Right(s) =>
              size = s
            case Left(iter) =>
              // Not enough space to unroll this block; drop to disk if applicable
              if (level.useDisk) {
                logWarning(s"Persisting block $blockId to disk instead.")
                diskStore.put(blockId) { channel =>
                  val out = Channels.newOutputStream(channel)
                  serializerManager.dataSerializeStream(blockId, out, iter)(classTag)
                }
                size = diskStore.getSize(blockId)
              } else {
                iteratorFromFailedMemoryStorePut = Some(iter)
              }
          }
        } else { // !level.deserialized
          memoryStore.putIteratorAsBytes(blockId, iterator(), classTag, level.memoryMode) match {
            case Right(s) =>
              size = s
            case Left(partiallySerializedValues) =>
              // Not enough space to unroll this block; drop to disk if applicable
              if (level.useDisk) {
                logWarning(s"Persisting block $blockId to disk instead.")
                diskStore.put(blockId) { channel =>
                  val out = Channels.newOutputStream(channel)
                  partiallySerializedValues.finishWritingToStream(out)
                }
                size = diskStore.getSize(blockId)
              } else {
                iteratorFromFailedMemoryStorePut = Some(partiallySerializedValues.valuesIterator)
              }
          }
        }
```

When StorageLevel has `deserialized=true` `putIteratorAsValues` is used and when `deserialized=false`, `putIteratorAsBytes` is used. Since data in python is always serialized as mentioned by @ScrapCodes, we should always use `putIteratorAsBytes`. To do this, we need to always keep StorageLevel with `deserialized=false`. Calling pyspark's `persist()` ensures that.

@cloud-fan @HyukjinKwon Can you please comment if the approach taken by this PR is correct?

Reply:
> Since data in python is always serialized

As I explained, this is not true in Spark SQL. The data to be cached is `RDD[CachedBatch]`. We should make sure `df.cache/persist` is consistent between python and scala/java.

Reply:
@cloud-fan @srowen 
Okay so, currently `scalaDataFrame.cache()` and `pySparkDataFrame.cache()` show consistent behavior. But `scalaDataFrame.persist()` and `pySparkDataFrame.persist()` are not consistent because `pysparkDataframe.persist()` uses `StorageLevel(true, true, false, false)` but `scalaDataframe.persist()` uses `StorageLevel(true, true, false, true)`. Should we change `pysparkDataframe.persist()` to directly call `self._jdf.persist()` ?

Reply:
Yea I think so

Reply:
okay i'll update the PR

Reply:
@cloud-fan as you explained, data is not serialized when performing SQL operations with pyspark. 

So should we revert changes done as part of [SPARK-12091](https://issues.apache.org/jira/browse/SPARK-12091) since users may need to set storage levels which have deserialized is equal to true?

Reply:
I'm not sure if it worth. The `CachedBatch` is mostly a wrapper of `byte[]` so keeping the serialized or deserialized form in memory doesn't matter too much.

I think the important thing is to keep Scala and Python `cache/persistent` consistent, although the Scala side has more builtin `StorageLevel`s.